tasks/add-unique-granuleID/src/index.ts

```diff
@@ -0,0 +1,57 @@
+import Logger from '@cumulus/logger';
+import { Context } from 'aws-lambda';
+
+import { constructCollectionId } from '@cumulus/message/Collections';
+import { generateUniqueGranuleId } from '@cumulus/ingest/granule';
+import { CumulusMessage, CumulusRemoteMessage } from '@cumulus/types/message';
+import { CumulusMessageWithAssignedPayload, runCumulusTask } from '@cumulus/cumulus-message-adapter-js';
+
+import { HandlerEvent, HandlerOutput } from './types';
+
+const log = new Logger({ sender: '@cumulus/add-unique-granule-id' });
+
+/**
+ * If no producerId is assigned,
+ * assign uniqueIds to the granules in the payload,
+ * preserving the original ID in 'producerGranuleId'.
+ *
+ * @param event - input from the message adapter
+ * @returns the uri of the pan
+ */
+async function assignUniqueIds(event: HandlerEvent): Promise<HandlerOutput> {
+  const { granules } = event.input;
+  for (const granule of granules) {
+    if (!granule.producerGranuleId) {
+      const collectionId = granule.collectionId
+        || constructCollectionId(granule.dataType, granule.version);
+      const newGranuleId = generateUniqueGranuleId(
+        granule.granuleId,
+        collectionId,
+        Number(event?.config?.hashDepth || 8)
+      );
+      granule.producerGranuleId = granule.granuleId;
+      granule.granuleId = newGranuleId;
+    }
+  }
+  const output = { granules };
+  log.debug('Granule output', output);
+  return output;
```

Comment:
I haven't messed with many task components. There's nothing else that needs done to update the Cumulus Message? The Lambda's output is sufficient?

Reply:
That's correct - the CMA configuration can be set such that this output goes to the default `payload` *or* when integrated into a workflow it can send the output so somewhere like `payload.granules`.